crates/control/src/role_assignment.rs

```diff
@@ -94,15 +94,36 @@ impl RoleAssignment {
             || primary_state == PrimaryState::Ready
             || primary_state == PrimaryState::Set
         {
+            let role_player_one = Role::Keeper;
+            let role_player_two = context.optional_roles.get(0).copied().unwrap_or_default();
+            let role_player_three = context.optional_roles.get(1).copied().unwrap_or_default();
+            let mut role_player_four = context.optional_roles.get(2).copied().unwrap_or_default();
+            let mut role_player_five = context.optional_roles.get(3).copied().unwrap_or_default();
+            let mut role_player_six = context.optional_roles.get(4).copied().unwrap_or_default();
+            let role_player_seven = Role::Striker;
+
+            if let Some(game_controller_state) = context.game_controller_state {
+                if let Some(_penalty) = game_controller_state.penalties.seven {
+                    role_player_six = Role::Striker;
+                    if let Some(_penalty) = game_controller_state.penalties.six {
+                        role_player_five = Role::Striker;
+                        if let Some(_penalty) = game_controller_state.penalties.five {
+                            role_player_four = Role::Striker;
+                        };
+                    };
+                }
+            }
```

Comment:
This could be expressed as a `find` operation through the penalties.
If we store the roles for each player in a `Players` container, both can just be indexed by the PlayerNumber.
`find` the player with the highest player number that does not have a penalty and assign the striker role to that player.

Reply:
Please have a look at my new adjustments and check whether this is what you meant.

Reply:
Thanks. Algorithmically it is what I mean, yes.
However, we should not store player data in arrays. That's what the [`Players`](https://github.com/HULKs/hulk/blob/main/crates/types/src/players.rs) container is for. It allows you to index the elements via a [`PlayerNumber`](https://github.com/HULKs/hulk/blob/main/crates/spl_network_messages/src/lib.rs#L35-L44) as well as iterating over all elements.
The penalties are already stored that way, you don't need to copy them into an array to work with them.

Reply:
Where exactly are the penalties stored this way?

Reply:
`game_controller_state.penalties`

Reply:
Please recheck

Reply:
I don't like the collecting into a vector just to iterate over it again. But we can't use `.rev` on the Players iterator directly (see #415).
I would suggest something like this:
```rust
if let Some(striker) = [PlayerNumber::Six, PlayerNumber::Five, PlayerNumber::Four]
    .into_iter()
    .find(|player| game_controller_state.penalties[*player].is_none())
{
    player_roles[striker] = Role::Striker;
};
```
We don't need to iterate over all players anyway, just the ones we want to consider as replacement strikers.
From those, we want to find the first (counting down from number six) that does not have a penalty and assign the striker role to that player.

Reply:
I inserted your code